# Lab | Data Structuring and Combining Data

## Challenge 1: Combining & Cleaning Data

In this challenge, we will be working with the customer data from an insurance company, as we did in the two previous labs. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

But this time, we got new data, which can be found in the following 2 CSV files located at the links below.

- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv

Note that you'll need to clean and format the new data.

Observation:
- One option is to first combine the three datasets and then apply the cleaning function to the new combined dataset
- Another option would be to read the clean file you saved in the previous lab, and just clean the two new files and concatenate the three clean datasets

In [609]:
# Your code goes here
import pandas as pd
df_1=pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv')
df_2=pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv')
df_3=pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv')
print(df_1.head(),'\n',df_2.head(),'\n', df_3.head())


  Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    Income  Monthly Premium Auto Number of Open Complaints     Policy Type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   Vehicle Class  Total Claim Amount  
0  Four-Door Car 

In [610]:
print(df_1.columns)
print(df_2.columns)
print(df_3.columns)

Index(['Customer', 'ST', 'GENDER', 'Education', 'Customer Lifetime Value',
       'Income', 'Monthly Premium Auto', 'Number of Open Complaints',
       'Policy Type', 'Vehicle Class', 'Total Claim Amount'],
      dtype='object')
Index(['Customer', 'ST', 'GENDER', 'Education', 'Customer Lifetime Value',
       'Income', 'Monthly Premium Auto', 'Number of Open Complaints',
       'Total Claim Amount', 'Policy Type', 'Vehicle Class'],
      dtype='object')
Index(['Customer', 'State', 'Customer Lifetime Value', 'Education', 'Gender',
       'Income', 'Monthly Premium Auto', 'Number of Open Complaints',
       'Policy Type', 'Total Claim Amount', 'Vehicle Class'],
      dtype='object')


In [611]:
df_3.rename(columns={"State": "ST", "Gender": "GENDER"}, inplace=True)
df_final = pd.concat([df_1, df_2, df_3], axis=0, ignore_index=True)
df_final.head()

Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    Income  Monthly Premium Auto Number of Open Complaints     Policy Type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   Vehicle Class  Total Claim Amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

In [612]:
df_copy=df_final.copy()
print(f'number of duplicates: {df_copy.duplicated().sum()}')  # Number of duplicates
df_copy = df_copy.drop_duplicates()
print(f'number of duplicates after cleaning:{df_copy.duplicated().sum()}')
print(df_copy.isnull().sum())



number of duplicates: 2939
number of duplicates after cleaning:0
Customer                       1
ST                             1
GENDER                       123
Education                      1
Customer Lifetime Value        8
Income                         1
Monthly Premium Auto           1
Number of Open Complaints      1
Policy Type                    1
Vehicle Class                  1
Total Claim Amount             1
dtype: int64


In [613]:
df_copy["GENDER"] = df_copy["GENDER"].replace({"female": "F","Male":"M","Femal":"F"})
df_copy["GENDER"]=df_copy["GENDER"].fillna("Unknown")
print(df_copy["GENDER"].unique())

['Unknown' 'F' 'M']


In [614]:
print(df_copy["Education"].unique())
df_copy["Education"] = df_copy["Education"].replace({"Bachelors": "Bachelor"})
print(df_copy["Education"].unique())

['Master' 'Bachelor' 'High School or Below' 'College' 'Bachelors' 'Doctor'
 nan]
['Master' 'Bachelor' 'High School or Below' 'College' 'Doctor' nan]


In [615]:
df_copy["Customer Lifetime Value"] = df_copy["Customer Lifetime Value"].str.replace("%", "").astype(float)
df_copy["Customer Lifetime Value"]=df_copy["Customer Lifetime Value"].fillna(df_copy["Customer Lifetime Value"].median())
df_copy["Customer Lifetime Value"] = df_copy["Customer Lifetime Value"].astype(int)
print(df_copy["Customer Lifetime Value"].unique())


[ 572076  697953 1288743 ...  568964  368672  399258]


In [616]:
df_copy["Number of Open Complaints"] = pd.to_numeric(df_copy["Number of Open Complaints"], errors="coerce")
df_copy["Number of Open Complaints"]=df_copy["Number of Open Complaints"].fillna(df_copy["Number of Open Complaints"].median())
df_copy["Number of Open Complaints"] = df_copy["Number of Open Complaints"].astype(int)
print(df_copy["Number of Open Complaints"].unique())


[0 2 3 1 5 4]


In [617]:
df_copy = df_copy.dropna(subset=["Policy Type"])
print(df_copy["Policy Type"].unique())

['Personal Auto' 'Corporate Auto' 'Special Auto']


In [618]:
print(df_copy.isnull().sum())


Customer                     0
ST                           0
GENDER                       0
Education                    0
Customer Lifetime Value      0
Income                       0
Monthly Premium Auto         0
Number of Open Complaints    0
Policy Type                  0
Vehicle Class                0
Total Claim Amount           0
dtype: int64


# Challenge 2: Structuring Data

In this challenge, we will continue to work with customer data from an insurance company, but we will use a dataset with more columns, called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by performing data cleaning, formatting, and structuring.

In [619]:
df_MCA=pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv')
df_MCA.head()


unnamed:_0 customer       state  customer_lifetime_value response  \
0           0  DK49336     Arizona              4809.216960       No   
1           1  KX64629  California              2228.525238       No   
2           2  LZ68649  Washington             14947.917300       No   
3           3  XL78013      Oregon             22332.439460      Yes   
4           4  QA50777      Oregon              9025.067525       No   

   coverage education effective_to_date employmentstatus gender  ...  \
0     Basic   College        2011-02-18         Employed      M  ...   
1     Basic   College        2011-01-18       Unemployed      F  ...   
2     Basic  Bachelor        2011-02-10         Employed      M  ...   
3  Extended   College        2011-01-11         Employed      M  ...   
4   Premium  Bachelor        2011-01-17    Medical Leave      F  ...   

   number_of_policies     policy_type        policy  renew_offer_type  \
0                   9  Corporate Auto  Corporate L3            Offer3   
1                   1   Personal Auto   Personal L3            Offer4   
2                   2   Personal Auto   Personal L3            Offer3   
3                   2  Corporate Auto  Corporate L3            Offer2   
4                   7   Personal Auto   Personal L2            Offer1   

   sales_channel  total_claim_amount  vehicle_class  vehicle_size  \
0          Agent          292.800000  Four-Door Car       Medsize   
1    Call Center          744.924331  Four-Door Car       Medsize   
2    Call Center          480.000000            SUV       Medsize   
3         Branch          484.013411  Four-Door Car       Medsize   
4         Branch          707.925645  Four-Door Car       Medsize   

  vehicle_type month  
0            A     2  
1            A     1  
2            A     2  
3            A     1  
4            A     1  

[5 rows x 27 columns]

In [620]:
df_MCA.drop(columns=["unnamed:_0"], inplace=True)
print(df_MCA.dtypes)

customer                          object
state                             object
customer_lifetime_value          float64
response                          object
coverage                          object
education                         object
effective_to_date                 object
employmentstatus                  object
gender                            object
income                             int64
location_code                     object
marital_status                    object
monthly_premium_auto               int64
months_since_last_claim          float64
months_since_policy_inception      int64
number_of_open_complaints        float64
number_of_policies                 int64
policy_type                       object
policy                            object
renew_offer_type                  object
sales_channel                     object
total_claim_amount               float64
vehicle_class                     object
vehicle_size                      object
vehicle_type    

1. You work at the marketing department and you want to know which sales channel brought the most sales in terms of total revenue. Using pivot, create a summary table showing the total revenue for each sales channel (branch, call center, web, and mail).
Round the total revenue to 2 decimal points.  Analyze the resulting table to draw insights.

In [621]:
# create a summary table showing the total revenue for each sales channel
sales_summary = df_MCA.pivot_table(
    values="total_claim_amount", 
    index="sales_channel", 
    aggfunc="sum"
).round(2)
print(sales_summary)

               total_claim_amount
sales_channel                    
Agent                  1810226.82
Branch                 1301204.00
Call Center             926600.82
Web                     706600.04


-  The Agent channel has generated the highest revenue (1,810,226.82)

- The Branch channel follows closely behind (1,301,204.00)

- While lower than Agent and Branch, Call Center sales (926,600.82) still represent a valuable channel

- Digital sales via Web (706,600.04) are the least performing channel

2. Create a pivot table that shows the average customer lifetime value per gender and education level. Analyze the resulting table to draw insights.

In [622]:
#Create a pivot table that shows the average customer lifetime value per gender and education level
CLV_summary = df_MCA.pivot_table(
    values="customer_lifetime_value", 
    index="gender", 
    columns="education", 
    aggfunc="mean"
).round(2)

# Display the results
print(CLV_summary)


education  Bachelor  College   Doctor  High School or Below   Master
gender                                                              
F           7874.27  7748.82  7328.51               8675.22  8157.05
M           7703.60  8052.46  7415.33               8149.69  8168.83


The highest CLV among females (8,675.22) is seen in the "High School or Below" category

Males in the same category also have a relatively high CLV (8,149.69)

Both genders show fairly similar CLV values across "Bachelor", "College", and "Master", hovering around 7,700 - 8,100

The "College" category differs slightly, with males having a higher CLV (8,052.46) than females (7,748.82).

The "Doctor" category has the lowest CLV across both genders (7328.51 for females, 7415.33 for males)


## Bonus

You work at the customer service department and you want to know which months had the highest number of complaints by policy type category. Create a summary table showing the number of complaints by policy type and month.
Show it in a long format table.

*In data analysis, a long format table is a way of structuring data in which each observation or measurement is stored in a separate row of the table. The key characteristic of a long format table is that each column represents a single variable, and each row represents a single observation of that variable.*

*More information about long and wide format tables here: https://www.statology.org/long-vs-wide-data/*

In [623]:

# Creating the pivot table in long format
complaints_summary = df_MCA.pivot_table(
    values="number_of_open_complaints", 
    index=["policy_type", "month"],  
    aggfunc="count"
).reset_index()

# Display the result
print(complaints_summary)

      policy_type  month  number_of_open_complaints
0  Corporate Auto      1                       1252
1  Corporate Auto      2                       1089
2   Personal Auto      1                       4329
3   Personal Auto      2                       3799
4    Special Auto      1                        237
5    Special Auto      2                        204
